Adding warning to exceptions that describes field conflicts and mapping issues (#3025)
elastic/security-docs: docs/troubleshooting/detections/detection-rules.asciidoc

CAUTION: Turning off `autocomplete:useTimeRange` could cause performance issues if the data set is especially large.
====

[discrete]
[[rule-exceptions-field-conflicts]]
.Warning about type conflicts and unmapped fields 
[%collapsible]
====

> A warning icon (image:images/field-warning-icon.png[Field conflict warning icon,13,13]) and message appear for fields that have <<fields-with-conflicting-types,type conflicts>> or are <<unmapped-field-conflict,unmapped>> across multiple indices. You can learn more about the conflict by hovering over the field. After you select it, the warning message is displayed beneath the field.

Reviewer: ```suggestion
A warning icon (image:images/field-warning-icon.png[Field conflict warning icon,13,13]) and message appear for fields with <<fields-with-conflicting-types,type conflicts>> and <<unmapped-field-conflict,unmapped>> fields. You can learn more about the problem by hovering over the field. After you select it, the warning message appears beneath the field.
```
I could be wrong but I don't think "across multiple indices" applies to unmapped fields, right? They just aren't mapped within a particular index, if I understand correctly.

Reviewer 2: I actually prefer `conflict` vs. `problem` here.

Author: @benironside your understanding is correct for unmapped fields (i.e., multiple fields can be unmapped in multiple indices) but "across multiple indices" still applies for type conflicts because a field can be defined as different types in multiple indices. 

Also -- and @e40pud correct me if I'm wrong -- my understanding is the warning appears if one or both of the conditions are met (i.e., when a field has type conflicts, is unmapped, or both) so "or" is more appropriate than "and" in the first sentence. For example:

```suggestion
A warning icon (image:images/field-warning-icon.png[Field conflict warning icon,13,13]) and message appear for fields with <<fields-with-conflicting-types,type conflicts>> across multiple indices or are <<unmapped-field-conflict,unmapped fields>>. You can learn more about the conflict by hovering over the field. After you select it, the warning message appears beneath the field.
```

> A warning icon (image:images/field-warning-icon.png[Field conflict warning icon,15,15]) and message appear for fields that have <<fields-with-conflicting-types,type conflicts>> or are <<unmapped-field-conflict,unmapped>> across specified indices. You can learn more about the conflict by hovering over the field. After you select it, the warning message is displayed beneath the field.

Author: ```suggestion
A warning icon (image:images/field-warning-icon.png[Field conflict warning icon,13,13]) and message appear for fields that have <<fields-with-conflicting-types,type conflicts>> or are <<unmapped-field-conflict,unmapped>> across specified indices. You can learn more about the conflict by hovering over the field. After you select it, the warning message is displayed beneath the field.
```

> A warning icon (image:images/field-warning-icon.png[Field conflict warning icon,13,13]) and message appear for fields that have <<fields-with-conflicting-types,type conflicts>> or are <<unmapped-field-conflict,unmapped>> across specified indices. You can learn more about the conflict by hovering over the field. After you select it, the warning message is displayed beneath the field.

Author: ```suggestion
A warning icon (image:images/field-warning-icon.png[Field conflict warning icon,13,13]) and message appear for fields that have <<fields-with-conflicting-types,type conflicts>> or are <<unmapped-field-conflict,unmapped>> across multiple indices. You can learn more about the conflict by hovering over the field. After you select it, the warning message is displayed beneath the field.
```

> A warning icon (image:images/field-warning-icon.png[Field conflict warning icon,15,15]) and message appears for fields that have <<fields-with-conflicting-types,conflicting types>> or are <<unmapped-field-conflict,unmapped>> across specified indices. You can learn more about the conflict by hovering over the field. After you select it, the warning message is displayed beneath the field.

Author: ```suggestion
A warning icon (image:images/field-warning-icon.png[Field conflict warning icon,15,15]) and message appear for fields that have <<fields-with-conflicting-types,type conflicts>> or are <<unmapped-field-conflict,unmapped>> across specified indices. You can learn more about the conflict by hovering over the field. After you select it, the warning message is displayed beneath the field.
```

> NOTE: A field can have conflicting types _and_ be unmapped in specified indices.  

Author: ```suggestion
NOTE: A field can have type conflicts _and_ be unmapped in specified indices.  
```

[role="screenshot"]
image::images/warning-icon-message.png[Shows the warning icon and message,70%]

[float]
[[fields-with-conflicting-types]]
==== Fields with conflicting types  

> Type conflicts occur when a field is mapped to different types across multiple indices. To resolve this issue, consider creating new indices with the correct type mappings and {ref}/docs-reindex.html[reindexing your data]. Otherwise, use the information about a field's type mapping to ensure you're entering correct field values when defining exception conditions. 

Reviewer: ```suggestion
Type conflicts occur when a field is mapped to different types across multiple indices. To resolve this issue, you can create new indices with matching field type mappings and {ref}/docs-reindex.html[reindex your data]. Otherwise, use the information about a field's type mappings to ensure you're entering compatible field values when defining exception conditions. 
```

In the following example, the selected field has been defined as different types across five indices.

[role="screenshot"]

> image::images/warning-type-conflicts.png[Warning for unmapped fields,80%] 

Author: ```suggestion
image::images/warning-type-conflicts.png[Warning for unmapped fields,70%] 
```

[float]
[[unmapped-field-conflict]]
==== Unmapped fields 

Unmapped fields are undefined within an index's mapping definition. Selecting an unmapped field could stop the exception from being applied to the rule's indices. This could lead to false positives or unexpected alerts being created.  

In the following example, the selected field has type conflicts and is unmapped across two indices. It also has type conflicts in three indices.

[role="screenshot"]

> image::images/warning-unmapped-fields.png[Warning for unmapped fields,80%] 

Author: ```suggestion
image::images/warning-unmapped-fields.png[Warning for unmapped fields,70%] 
```

> Unmapped fields are undefined within an index's mapping definition. Selecting an unmapped field could prevent the exception from being applied to a rule's indices and lead to false positives or unexpected alerts. To fix unmapped fields, {ref}/explicit-mapping.html#update-mapping[add them] to your indices' mapping definitions. 

Reviewer: ```suggestion
Unmapped fields are undefined within an index's mapping definition. Using unmapped fields to define an exception can prevent it from working as expected, and lead to false positives or unexpected alerts. To fix unmapped fields, {ref}/explicit-mapping.html#update-mapping[add them] to your indices' mapping definitions. 
```
